File: api/webai/v1/chat/completions.py
# WebAI-to-API (g4f) wrapper for Vercel
import json
import os
from pathlib import Path
import sys

# Add the WebAI-to-API package to the Python path
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'packages' / 'WebAI-to-API' / 'src'))

def handler(request, response):
    if request.method != 'POST':
        response.status = 405
        response.send(json.dumps({'error': 'Method not allowed'}))
        return

    try:
        body = json.loads(request.body)
        messages = body.get('messages', [])
        model = body.get('model', 'gpt-3.5-turbo')
        stream = body.get('stream', False)

        if not messages:
            response.status = 400
            response.send(json.dumps({'error': 'Missing messages'}))
            return

        # Try to use g4f if available
        try:
            import g4f
            from g4f.client import Client as G4FClient
            
            # Initialize g4f client
            client = G4FClient()
            
            # Convert messages format if needed
            chat_messages = []
            for msg in messages:
                chat_messages.append({
                    'role': msg.get('role', 'user'),
                    'content': msg.get('content', '')
                })
            
            # Make request to g4f
            g4f_response = client.chat.completions.create(
                model=model,
                messages=chat_messages,
                stream=stream
            )
            
            if stream:
                # Handle streaming response
                response.status = 200
                response.headers['Content-Type'] = 'text/event-stream'
                response.headers['Access-Control-Allow-Origin'] = '*'
                
                # For simplicity, collect all chunks and send as one response
                # In production, you'd want to stream properly
                full_content = ""
                for chunk in g4f_response:
                    if hasattr(chunk, 'choices') and chunk.choices:
                        delta = chunk.choices[0].delta
                        if hasattr(delta, 'content') and delta.content:
                            full_content += delta.content
                
                # Send as non-streaming response for Vercel compatibility
                response.send(json.dumps({
                    'id': f'chatcmpl-webai-{int(os.time())}',
                    'object': 'chat.completion',
                    'created': int(os.time()),
                    'model': model,
                    'choices': [{
                        'index': 0,
                        'message': {
                            'role': 'assistant',
                            'content': full_content
                        },
                        'finish_reason': 'stop'
                    }],
                    'usage': {
                        'prompt_tokens': 0,
                        'completion_tokens': 0,
                        'total_tokens': 0
                    }
                }))
            else:
                # Non-streaming response
                content = ""
                if hasattr(g4f_response, 'choices') and g4f_response.choices:
                    content = g4f_response.choices[0].message.content
                elif hasattr(g4f_response, 'content'):
                    content = g4f_response.content
                else:
                    content = str(g4f_response)
                
                response.status = 200
                response.send(json.dumps({
                    'id': f'chatcmpl-webai-{int(os.time())}',
                    'object': 'chat.completion',
                    'created': int(os.time()),
                    'model': model,
                    'choices': [{
                        'index': 0,
                        'message': {
                            'role': 'assistant',
                            'content': content
                        },
                        'finish_reason': 'stop'
                    }],
                    'usage': {
                        'prompt_tokens': 0,
                        'completion_tokens': 0,
                        'total_tokens': 0
                    }
                }))
                
        except ImportError:
            # Fallback: Use WebAI-to-API's FastAPI app if g4f is not available
            # This would require running the FastAPI app separately or using httpx to call it
            response.status = 503
            response.send(json.dumps({
                'error': 'g4f library not available. Please install g4f in the Python environment.',
                'details': 'WebAI-to-API requires g4f library to be installed.'
            }))
            
    except Exception as e:
        print(f"WebAI-to-API Error: {e}")
        response.status = 500
        response.send(json.dumps({'error': 'WebAI-to-API generation failed', 'details': str(e)}))

# Vercel Python runtime expects a 'handler' function
if __name__ == '__main__':
    class MockRequest:
        def __init__(self, method, headers, body):
            self.method = method
            self.headers = headers
            self.body = body

    class MockResponse:
        def __init__(self):
            self.status = 200
            self._data = None
            self.headers = {}

        def send(self, data):
            self._data = data

        def json(self):
            return json.loads(self._data)

    # Example usage for local testing
    mock_req = MockRequest(
        method='POST',
        headers={'Content-Type': 'application/json'},
        body=json.dumps({
            'model': 'gpt-3.5-turbo',
            'messages': [{'role': 'user', 'content': 'Hello, how are you?'}]
        })
    )
    mock_res = MockResponse()
    handler(mock_req, mock_res)
    print(mock_res.json())

File: api/webai/v1/chat/completions.js
import { NextResponse } from 'next/server';
import { ChatCompletion } from 'g4f';

export const config = {
  runtime: 'nodejs',
};

export default async function handler(req) {
  // Handle OPTIONS
  if (req.method === 'OPTIONS') {
    return new NextResponse(null, {
      status: 200,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Access-Control-Allow-Headers': 'Content-Type, Authorization',
      },
    });
  }

  // Only allow POST
  if (req.method !== 'POST') {
    return NextResponse.json(
      { error: 'Method not allowed', details: 'Only POST requests are supported' },
      { status: 405, headers: { 'Access-Control-Allow-Origin': '*' } }
    );
  }

  try {
    const body = await req.json();
    const { model, messages, stream = false } = body;

    if (!model || !messages || !Array.isArray(messages) || messages.length === 0) {
      return NextResponse.json(
        { error: 'Bad request', details: 'Model and messages array are required' },
        { status: 400, headers: { 'Access-Control-Allow-Origin': '*' } }
      );
    }

    // Create timeout
    const controller = new AbortController();
    const timeoutId = setTimeout(() => {
      controller.abort();
    }, 30000); // 30 second timeout

    try {
      // Use g4f-ts to generate response
      const chatCompletion = new ChatCompletion();
      
      // Convert messages format - g4f-ts expects OpenAI format
      const conversation = messages.map(msg => ({
        role: msg.role,
        content: msg.content,
      }));

      // Generate response with timeout
      const timeoutPromise = new Promise((_, reject) => {
        setTimeout(() => reject(new Error('Request timeout')), 30000);
      });

      const responsePromise = chatCompletion.create({
        model: model,
        messages: conversation,
        stream: stream,
      });

      // Race between response and timeout
      const response = await Promise.race([
        responsePromise,
        timeoutPromise
      ]);

      clearTimeout(timeoutId);

      // Handle streaming vs non-streaming
      if (stream) {
        // For streaming, we need to handle it differently
        // For now, collect all chunks
        let fullContent = '';
        if (response && typeof response === 'object' && 'next' in response) {
          // It's an async iterator
          for await (const chunk of response) {
            if (chunk.choices && chunk.choices[0]?.delta?.content) {
              fullContent += chunk.choices[0].delta.content;
            }
          }
        } else if (typeof response === 'string') {
          fullContent = response;
        } else if (response?.choices?.[0]?.message?.content) {
          fullContent = response.choices[0].message.content;
        }

        // Return non-streaming response for simplicity
        return NextResponse.json({
          id: `chatcmpl-${Date.now()}`,
          object: 'chat.completion',
          created: Math.floor(Date.now() / 1000),
          model: model,
          choices: [{
            index: 0,
            message: {
              role: 'assistant',
              content: fullContent,
            },
            finish_reason: 'stop',
          }],
          usage: {
            prompt_tokens: messages.reduce((sum, m) => sum + (m.content?.length || 0), 0) / 4,
            completion_tokens: fullContent.length / 4,
            total_tokens: (messages.reduce((sum, m) => sum + (m.content?.length || 0), 0) + fullContent.length) / 4,
          },
        }, {
          status: 200,
          headers: {
            'Access-Control-Allow-Origin': '*',
          },
        });
      } else {
        // Non-streaming response
        let content = '';
        
        if (typeof response === 'string') {
          content = response;
        } else if (response?.choices?.[0]?.message?.content) {
          content = response.choices[0].message.content;
        } else if (response?.content) {
          content = response.content;
        } else {
          // Try to extract content from various possible formats
          content = JSON.stringify(response);
        }

        if (!content) {
          throw new Error('No content in response');
        }

        return NextResponse.json({
          id: `chatcmpl-${Date.now()}`,
          object: 'chat.completion',
          created: Math.floor(Date.now() / 1000),
          model: model,
          choices: [{
            index: 0,
            message: {
              role: 'assistant',
              content: content,
            },
            finish_reason: 'stop',
          }],
          usage: {
            prompt_tokens: messages.reduce((sum, m) => sum + (m.content?.length || 0), 0) / 4,
            completion_tokens: content.length / 4,
            total_tokens: (messages.reduce((sum, m) => sum + (m.content?.length || 0), 0) + content.length) / 4,
          },
        }, {
          status: 200,
          headers: {
            'Access-Control-Allow-Origin': '*',
          },
        });
      }
    } catch (fetchError) {
      clearTimeout(timeoutId);

      // Handle timeout
      if (fetchError.name === 'AbortError' || fetchError.message?.includes('timeout') || fetchError.message?.includes('aborted')) {
        console.error(`[WebAI] Request timeout for model: ${model}`);
        return NextResponse.json(
          {
            error: 'Request timeout',
            details: 'The request took too long to complete (over 30 seconds). Please try again with a different model.',
            model: model,
          },
          {
            status: 504,
            headers: { 'Access-Control-Allow-Origin': '*' }
          }
        );
      }

      // Handle other errors
      console.error(`[WebAI] Error for model: ${model}`, fetchError);
      return NextResponse.json(
        {
          error: 'Model error',
          details: fetchError.message || 'Failed to generate response',
          model: model,
          errorType: fetchError.name || 'UnknownError',
        },
        {
          status: 500,
          headers: { 'Access-Control-Allow-Origin': '*' }
        }
      );
    }
  } catch (error) {
    console.error('[WebAI] Top-level error:', error);
    return NextResponse.json(
      {
        error: 'Internal Server Error',
        details: error.message || 'An unexpected error occurred',
        errorType: error.name || 'UnknownError',
      },
      {
        status: 500,
        headers: { 'Access-Control-Allow-Origin': '*' }
      }
    );
  }
}

